lib/Sema/CSApply.cpp

```diff
@@ -1615,6 +1620,66 @@ namespace {
         return forceUnwrapIfExpected(ref, memberLocator);
       }
 
+      if (requiresDistributedThunk(base, memberLoc.getStartLoc(), memberRef)) {
+        auto *decl = memberRef.getDecl();
+        FuncDecl *thunkDecl = nullptr;
+        if (auto *FD = dyn_cast<FuncDecl>(decl)) {
+          thunkDecl = FD->getDistributedThunk();
+        } else {
+          thunkDecl = cast<VarDecl>(decl)->getDistributedThunk();
+        }
+
+        if (!thunkDecl)
+          return nullptr;
+
+        auto thunkType = refTy;
+        auto thunkOpenedType = openedType;
+
+        // If this is a reference to a computed property then we need to
+        // form a function type from it with unapplied Self so
+        // (Self) -> T becomes (Self) -> () -> T
+        if (isa<VarDecl>(decl)) {
+          auto extInfo = ASTExtInfoBuilder().withAsync().withThrows().build();
+
+          thunkOpenedType =
+              FunctionType::get(/*params=*/{}, thunkOpenedType, extInfo);
+          thunkType =
+              FunctionType::get({FunctionType::Param(selfTy)}, thunkOpenedType);
+        }
+
+        ConcreteDeclRef thunkRef{thunkDecl, memberRef.getSubstitutions()};
+
+        auto declRefExpr = new (context) DeclRefExpr(
+            thunkRef, memberLoc, Implicit, AccessSemantics::DirectToStorage);
+
+        declRefExpr->setFunctionRefKind(choice.getFunctionRefKind());
+        declRefExpr->setType(thunkType);
+
+        cs.cacheType(declRefExpr);
```

Comment:
oh that's new to me, why do we do this?

Reply:
Are you talking about `cacheType` specifically? That's used to let `finishApply` know what the type was for the newly created node, since all the expressions involved in solving should have their types cached.

Reply:
Yeah, that -- I see, thanks :)